rocksdb/rocksdb.nim

```diff
@@ -0,0 +1,251 @@
+
+  if columnFamilies.len == 0:
+    return err("rocksdb: no column families")
+
+  var
+    cfNames = columnFamilies.mapIt(it.name().cstring)
+    cfOpts = columnFamilies.mapIt(it.options.cPtr)
+    columnFamilyHandles = newSeq[ColFamilyHandlePtr](columnFamilies.len)
+    errors: cstring
+
+  let rocksDbPtr = rocksdb_open_column_families(
+        dbOpts.cPtr,
+        path.cstring,
+        cfNames.len().cint,
+        cast[cstringArray](cfNames[0].addr),
+        cfOpts[0].addr,
+        columnFamilyHandles[0].addr,
+        cast[cstringArray](errors.addr))
+  bailOnErrors(errors)
+
+  var cfTable = newColFamilyTable()
+  for i, cf in columnFamilies:
+    cfTable.put(cf.name(), columnFamilyHandles[i])
+
+  let db = RocksDbReadWriteRef(
+      cPtr: rocksDbPtr,
+      path: path,
+      dbOpts: dbOpts,
+      readOpts: readOpts,
+      writeOpts: writeOpts,
+      cfTable: cfTable)
+  ok(db)
+
+proc openRocksDbReadOnly*(
+    path: string,
+    dbOpts = defaultDbOptions(),
+    readOpts = defaultReadOptions(),
+    columnFamilies = @[defaultColFamilyDescriptor()],
+    errorIfWalFileExists = false): RocksDBResult[RocksDbReadOnlyRef] =
+
+  if columnFamilies.len == 0:
+    return err("rocksdb: no column families")
+
+  var
+    cfNames = columnFamilies.mapIt(it.name().cstring)
+    cfOpts = columnFamilies.mapIt(it.options.cPtr)
+    columnFamilyHandles = newSeq[ColFamilyHandlePtr](columnFamilies.len)
+    errors: cstring
+
+  let rocksDbPtr = rocksdb_open_for_read_only_column_families(
+        dbOpts.cPtr,
+        path.cstring,
+        cfNames.len().cint,
+        cast[cstringArray](cfNames[0].addr),
+        cfOpts[0].addr,
+        columnFamilyHandles[0].addr,
+        errorIfWalFileExists.uint8,
+        cast[cstringArray](errors.addr))
+  bailOnErrors(errors)
+
+  var cfTable = newColFamilyTable()
+  for i, cf in columnFamilies:
+    cfTable.put(cf.name(), columnFamilyHandles[i])
+
+  let db = RocksDbReadOnlyRef(
+      cPtr: rocksDbPtr,
+      path: path,
+      dbOpts: dbOpts,
+      readOpts: readOpts,
+      cfTable: cfTable)
+  ok(db)
+
+template isClosed*(db: RocksDbRef): bool =
+  db.cPtr.isNil()
+
+proc cPtr*(db: RocksDbRef): RocksDbPtr =
+  doAssert not db.isClosed()
+  db.cPtr
+
+proc get*(
+    db: RocksDbRef,
+    key: openArray[byte],
+    onData: DataProc,
+    columnFamily = "default"): RocksDBResult[bool] =
+
+  if key.len() == 0:
+    return err("rocksdb: key is empty")
+
+  let cfHandle = db.cfTable.get(columnFamily)
+  if cfHandle.isNil:
+    return err("rocksdb: unknown column family")
+
+  var
+    len: csize_t
+    errors: cstring
+  let data = rocksdb_get_cf(
+        db.cPtr,
+        db.readOpts.cPtr,
+        cfHandle.cPtr,
+        cast[cstring](unsafeAddr key[0]),
+        csize_t(key.len),
+        len.addr,
+        cast[cstringArray](errors.addr))
+  bailOnErrors(errors)
+
+  if data.isNil:
+    doAssert len == 0
+    ok(false)
+  else:
+    onData(toOpenArrayByte(data, 0, len.int - 1))
```

Comment:
My understanding is that toOpenArrayByte doesn't copy the data and passing an array into a proc openArray parameter doesn't copy either. But turning the openArray into a seq still requires at least one copy. I'm wondering if we can further reduce the number of copies by using the sink and/or lent constructs?